core-bundle/src/Resources/contao/modules/ModuleTwoFactor.php

```diff
@@ -63,12 +74,37 @@ protected function compile()
 			$this->disableTwoFactor($user, $return);
 		}
 
-		$this->Template->isEnabled = (bool) $this->User->useTwoFactor;
+		if (Input::post('FORM_SUBMIT') == 'tl_two_factor_show_backup_codes')
+		{
+			if (!$user->backupCodes || !\count(json_decode($user->backupCodes, true)))
+			{
+				$this->generateBackupCodes($user);
+			}
+
+			$this->Template->showBackupCodes = true;
+		}
+
+		if (Input::post('FORM_SUBMIT') == 'tl_two_factor_generate_backup_codes')
+		{
+			$this->generateBackupCodes($user);
+
+			$this->Template->showBackupCodes = true;
+		}
+
+		$this->Template->isEnabled = (bool) $user->useTwoFactor;
 		$this->Template->twoFactor = $GLOBALS['TL_LANG']['MSC']['twoFactorAuthentication'];
 		$this->Template->explain = $GLOBALS['TL_LANG']['MSC']['twoFactorExplain'];
 		$this->Template->active = $GLOBALS['TL_LANG']['MSC']['twoFactorActive'];
 		$this->Template->enableButton = $GLOBALS['TL_LANG']['MSC']['enable'];
 		$this->Template->disableButton = $GLOBALS['TL_LANG']['MSC']['disable'];
+		$this->Template->backupCodesLabel = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesLabel'];
+		$this->Template->backupCodesShow = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesShow'];
+		$this->Template->backupCodesExplain = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesExplain'];
+		$this->Template->backupCodesInfo = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesInfo'];
+		$this->Template->backupCodesGenerate = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesGenerate'];
+		$this->Template->backupCodesRegenerate = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesRegenerate'];
+		$this->Template->backupCodesRegenerateInfo = $GLOBALS['TL_LANG']['MSC']['twoFactorBackupCodesRegenerateInfo'];
```

Comment:
According to our discussion, we should use $this->trans() for labels in the template

Reply:
Changed in a22ba09d3ac32e77bc78600b373a50e8bb283dc3.